burr/core/application.py

```diff
@@ -1127,38 +1181,152 @@ def with_hooks(self, *adapters: LifecycleAdapter) -> "ApplicationBuilder":
+        return self
+
+    def initialize_from(
+        self,
+        persister: BasicStatePersistence,
+        resume_at_next_action: bool,
+        default_state: dict,
+        default_entry_point: str,
+    ) -> "ApplicationBuilder":
+        self.persister = persister
+        self.resume_at_next_action = resume_at_next_action
+        self.default_state = default_state
+        self.start = default_entry_point
+        return self
+
+    def with_state_persister(
+        self, persister: BasicStatePersistence, on_every: str = "step"
+    ) -> "ApplicationBuilder":
+        """Adds a state persister to the application. This is a way to persist state out to a database, file, etc...
+        at the specified interval.
+
+        :param persister: The persister to add
+        :param on_every: The interval to persist state. Currently only "step" is supported.
+        :return: The application builder for future chaining.
+        """
+        if on_every != "step":
+            raise ValueError(f"on_every {on_every} not supported")
+        # stick persister into a post_step_hook if it doesn't implement it already.
+        if hasattr(persister, "post_step_hook"):
+            if persister not in self.lifecycle_adapters:
+                self.lifecycle_adapters.append(persister)
         else:
-            raise ValueError(f"Tracker {tracker} not supported")
+            # TODO: pull this out more formally
+            class persisterWrapper(PostRunStepHook):
+                def post_run_step(
+                    self,
+                    app_id: str,
+                    partition_key: str,
+                    sequence_id: int,
+                    state: "State",
+                    action: "Action",
+                    result: Optional[Dict[str, Any]],
+                    exception: Exception,
+                    **future_kwargs: Any,
+                ):
+                    if exception is None:
+                        persister.save(
+                            partition_key, app_id, sequence_id, action.name, state, "completed"
+                        )
+                    else:
+                        persister.save(
+                            partition_key, app_id, sequence_id, action.name, state, "failed"
+                        )
+
+            self.lifecycle_adapters.append(persisterWrapper())
         return self
 
+    def _load_from_persister(self):
+        """Loads from the set persister and into this current object.
+
+        Mutates:
+         - self.state
+         - self.sequence_id
+         - maybe self.start
+
+        """
+        # load state from persister
+        load_result = self.persister.load(self.partition_key, self.app_id, self.sequence_id)
+        if load_result is None:
+            # there was nothing to load -- use default state
+            self.state = self.state.update(**self.default_state)
+            self.sequence_id = 0
```

Comment:
This will get it out of sync with the `increment_sequence_id` call... We should just have the `sequence_id` be a property that gets the state that's never set directly, and have a getter on it